nfd: init at 0.7.1

bertof opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Package the named Data Networking Forwarder Daemon (NFD)

###### Things done
- Add NFD derivation
- Link NFD derivation as a top-level package

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

bertof committed:
nfd: init at version 0.7.1

- Add NFD derivation
- Link NFD derivation as a top-level package

r-burns commented on pkgs/servers/nfd/default.nix:
```diff
@@ -0,0 +1,49 @@
+, libpcap
+, fetchFromGitHub
+, boost
+, pkgconfig
```
```suggestion
, pkg-config
```
pkgconfig is a compatibility alias (an artifact from before the nix language supported hyphens in names), pkg-config is preferred

SuperSandro2000 commented on pkgs/servers/nfd/default.nix:
```diff
@@ -0,0 +1,49 @@
+, systemd
+, python3
+, python3Packages
+, withSystemd ? true
```
```suggestion
, withSystemd ? stdenv.isLinux
```

SuperSandro2000 commented on pkgs/servers/nfd/default.nix:
```diff
@@ -0,0 +1,49 @@
+, python3Packages
+, withSystemd ? true
+, withWebSocket ? true
+}:
+let
+  pname = "nfd";
+  version = "0.7.1";
+  pythonPackages = p: with p; [ sphinx ];
+in
+stdenv.mkDerivation {
+  inherit pname version;
```
```suggestion
}:
let
  pythonPackages = p: with p; [ sphinx ];
in
stdenv.mkDerivation {
  pname = "nfd";
  version = "0.7.1";
```

SuperSandro2000 commented on pkgs/servers/nfd/default.nix:
```diff
@@ -0,0 +1,49 @@
+    "--boost-libs=${boost.out}/lib"
+  ] ++ lib.optional (!withWebSocket) "--without-websocket";
+
+  outputs = [ "out" "dev" ];
+  meta = with lib; {
```
```suggestion
  outputs = [ "out" "dev" ];

  meta = with lib; {
```

SuperSandro2000 commented on pkgs/servers/nfd/default.nix:
```diff
@@ -0,0 +1,49 @@
+    repo = lib.toUpper pname;
+    rev = "${lib.toUpper pname}-${version}";
+    sha256 = "1l9bchj8c68r6qw4vr1kc96jgxl0vpqa2vjkvy1xmhz92sivr6gi";
+    fetchSubmodules = withWebSocket;
```
Please use the websocketpp from nixpkgs.

bertof replied:
Websocketpp is 0.8.1 in nixpkgs while this project uses 0.8.1-hotfix. The [changelist](https://github.com/cawka/websocketpp/tags) shows the hotfix is to support boost 17.

bertof replied:
I just noticed boost is an alias for boost169, and not boost175.
In the [release notes](https://named-data.net/doc/NFD/current/RELEASE_NOTES.html) the authors suggest at least 1.6.5.
What is the common practice in this case?

bertof replied:
I tested it with nixpkgs websocketpp and works ok

bertof committed:
Applied revision suggestions

bertof committed:
Add fetchSubmodules

NFD uses websocket++ 0.8.1-hotfix in their stable release so that it can be built with boost 175.
ndn-cxx is built with boost 169 in nixpkgs, although it could use the newest version, so I'm matching the same version.

bertof committed:
Fixed tests and cleanup

bertof committed:
Fix typo in repo revision

bertof committed:
Add fetch submodules

NFD build script requires the complete websocket++ source in the build directory.
I could not build it reusing the packages websocket++.

r-burns commented on pkgs/servers/nfd/default.nix:
```diff
@@ -0,0 +1,53 @@
+    repo = lib.toUpper pname;
+    rev = "NFD-${version}";
+    sha256 = "1l9bchj8c68r6qw4vr1kc96jgxl0vpqa2vjkvy1xmhz92sivr6gi";
+    fetchSubmodules = withWebSocket;
```
I don't think this will work the way you're intending - withWebSocket doesn't change the sha256 so the submodules aren't being turned on/off here. I would probably just stick to fetching submodules unconditionally so you don't need complicated logic to change the sha256.

bertof replied:
You're right, the hash would change. I'll make it fetch the submodules unconditionally.

r-burns commented on pkgs/servers/nfd/default.nix:
```diff
@@ -0,0 +1,53 @@
+    "--boost-includes=${boost.dev}/include"
+    "--boost-libs=${boost.out}/lib"
+    "--with-tests"
+  ] ++ lib.optional (!withWebSocket) "--without-websocket";
```
```suggestion
    (lib.withFeature withWebSocket "websocket")
  ];
```

Providing the flag unconditionally avoids any worries about whether websocket is enabled or disabled by default by the configuration

bertof replied:
I didn't know of the `withFeature` function. Nice to have, I'll update the code.

bertof replied:
I just checked for this suggestion and unfortunately the build script won't work with a `--with-websocket` argument: it only checks for the without version and defaults to including it in the build.
A more pleasant version of the solution I hardcoded is with `lib.optionalString`

r-burns commented on pkgs/servers/nfd/default.nix:
```diff
@@ -0,0 +1,53 @@
+  ] ++ lib.optional (!withWebSocket) "--without-websocket";
+
+  doCheck = true;
+  checkPhase = ''
+    build/unit-tests-core
+    # build/unit-tests-daemon # 3 tests fail
+    build/unit-tests-rib
+    build/unit-tests-tools
+  '';
```
```suggestion
  checkPhase = ''
    runHook preCheck
    build/unit-tests-core
    # build/unit-tests-daemon # 3 tests fail
    build/unit-tests-rib
    build/unit-tests-tools
    runHook postCheck
  '';
```

Be sure to always run pre/post hooks when overriding phases

r-burns reviewed (commented):
Couple minor comments, looks good in general

bertof committed:
Call to pre and post check hooks

Co-authored-by: Ryan Burns <52847440+r-burns@users.noreply.github.com>

bertof committed:
Default to fetch NFD submodules

bertof commented:
I have a question about supporting MacOS. I'm building on NixOS and don't have access to a MacOS device.
From the NFD installation guide the build needs Xcode to be installed. How can I make sure it is in nix? I couldn't find an xcode derivation.

r-burns commented:
> I have a question about supporting MacOS. I'm building on NixOS and don't have access to a MacOS device.
From the NFD installation guide the build needs Xcode to be installed. How can I make sure it is in nix? I couldn't find an xcode derivation.

We don't have xcode for packaging purposes, I think because it's unfree so we couldn't use it on hydra. We do have a fake xcbuild package which provides some subset of functionality and is sometimes sufficient for packaging purposes.

I wouldn't worry about mac os if you don't have a machine to test on. Enterprising darwin users are free to make the necessary changes if they want to use it.

r-burns commented on pkgs/servers/nfd/default.nix:
```diff
@@ -0,0 +1,56 @@
+  meta = with lib; {
+    homepage = "http://named-data.net/";
+    description = "Named Data Neworking (NDN) Forwarding Daemon";
+    license = licenses.gpl3;
```
```suggestion
    license = licenses.gpl3Plus;
```

https://github.com/named-data/NFD/blob/master/COPYING.md

bertof committed:
Fix license

Co-authored-by: Ryan Burns <52847440+r-burns@users.noreply.github.com>

r-burns commented:
I was able to compile NFD, but I noticed that it's trying to load a config file from inside its store path:

```
NFD version 0.7.1 starting
Built with GNU C++ version 10.3.0, with GNU libstdc++ version 20210408, with Boost version 1.69.0, with libpcap version 1.10.1 (with TPACKET_V3), with WebSocket++ version 0.8.1, with ndn-cxx version 0.7.1
1636934920.642393 FATAL: [nfd.Main] ../daemon/common/config-file.cpp(88): Throw in function parse
Dynamic exception type: boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<nfd::ConfigFile::Error> >
std::exception::what: Failed to read configuration file /nix/store/yam2j6k12f3z9qzbj9c17dxbqrpbsdhw-nfd-0.7.1/etc/ndn/nfd.conf
===== Stacktrace =====
 0# 0x0000000000451832 in results/nfd/bin/nfd
 1# 0x0000000000582481 in results/nfd/bin/nfd
 2# 0x00000000005833BC in results/nfd/bin/nfd
 3# 0x0000000000480A4C in results/nfd/bin/nfd
 4# __libc_start_main in /nix/store/2zchy9mdx3kk166i98cx5sriq2c1bdsn-glibc-2.33-55/lib/libc.so.6
 5# 0x0000000000497D2A in results/nfd/bin/nfd
======================
```

Is it possible to configure it so that the default config file location is `/etc/ndn/nfd.conf`?

bertof commented:
> I was able to compile NFD, but I noticed that it's trying to load a config file from inside its store path 

Yes, it does search for `nfd.conf` or `nfd.conf.sample` by default.

> Is it possible to configure it so that the default config file location is `/etc/ndn/nfd.conf`?
> 
> Edit, I think you can do this by configuring with `--sysconfdir=/etc`

I'm not entirely sure it is possible without a patch but I'll give it a try

SuperSandro2000 commented on pkgs/servers/nfd/default.nix:
```diff
@@ -0,0 +1,56 @@
+  wafConfigureFlags = [
+    "--boost-includes=${boost.dev}/include"
+    "--boost-libs=${boost.out}/lib"
+    "--with-tests"
+    (lib.optionalString (!withWebSocket) "--without-websocket")
+  ];
```
```suggestion
    "--with-tests"
  ] + lib.optional (!withWebSocket) "--without-websocket";
```

bertof committed:
Switch to lib.optional for parameter string

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

bertof commented:
> Is it possible to configure it so that the default config file location is `/etc/ndn/nfd.conf`?
> Edit, I think you can do this by configuring with `--sysconfdir=/etc`

I checked and i can change the output directory of both binaries and configuration files to be inline with nix defaults by setting the `PREFIX` variable to `$out` through a `preConfigure` hook.

The authors, though, expect the user to configure the server manually first [copying the sample configuration](https://github.com/named-data/NFD/blob/master/docs/INSTALL.rst#general) as a starting point.

This is because in the default configuration [all security measures are disabled](https://github.com/named-data/NFD/blob/master/docs/INSTALL.rst#nfd-security) (for ease of test I suppose).

So in general you would always start it with `sudo nfd -c <path to custom configuration>`.
I agree that the error handling is quite rough and possibly misleading but it's apparently the intended behaviour.

I'm going to update the derivation with the custom `PREFIX` as per above.

bertof committed:
Change prefix as per nix defaults

r-burns commented:
Okay, thanks. I was thinking that the default config.sample would still be in the nix store, but nfd would look for its configuration at runtime in /etc like other nix packages. But I couldn't see a straightforward way to do this without patching, so requiring the config to be specified with -c seems reasonable.

I'm not sure what exporting PREFIX does, could you explain that last commit? Other than that this looks good to me :)

bertof commented:
You're once again right (I'm still getting used to all the nix build managers).
The `PREFIX` variable I'm referring to is used to set the [installation prefix](https://waf.io/apidocs/tools/gnu_dirs.html), and by default is `/usr/local`.
I thought I would have to set it to the derivation output folder to fix the path to the configuration files, but the `wafHook` already does that for us using the `--prefix` argument. 
My last change is redundant, I'll remove it.

After that everything should be fixed.

bertof committed:
Removed redundant prefix environment variable

r-burns reviewed (approved):
Great, thanks! Changes are looking good. If no one else reviews/merges in the next few days, feel free to give me a ping and I'll merge.

Result of `nixpkgs-review pr 144010` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>nfd</li>
  </ul>
</details>

bertof commented:
Thank you for your support! :smile:

bertof commented:
Hi! Seems like no other change or review have been proposed.
@r-burns or @SuperSandro2000 , would you like to merge?

SuperSandro2000 commented on pkgs/servers/nfd/default.nix:
```diff
@@ -0,0 +1,55 @@
+  '';
+
+  meta = with lib; {
+    homepage = "http://named-data.net/";
```
```suggestion
    homepage = "https://named-data.net/";
```

bertof committed:
Homepage to HTTPS

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

r-burns commented:
Thanks!
